### ISSUE TITLE
Feature request: `extract_row` and `extract_col` aliases

### ISSUE THREAD

**Issue by danklotz** [tidyverse/magrittr#134]
When I want to use the `extract` function to get chosen columns (or rows)  from data I currently use the alias in the following way:

```
data %>% extract( ,1:3) # and 
data %>% extract(1:3, )
```

Now, i personaly think that the extract(x , ) syntax looks a bit strange and Rstudio always marks it as bad syntax. Therefore I would propose two further aliases, i.e. `extract_row` and `extract_col` would be nice. So: 

```
 data %>% extract_col( 1:3) # == data %>% extract(  ,1:3)
data %>% extract_row(1:3) # data %>% extract(1:3, )
```

**Comment by Deleetdk** [comment]
Seems easier to just ask RStudio to exclude the function from their syntax checker.

You only need to leave an empty argument slot when you want rows:

```
#rows
data %>% extract(1:3, )
#cols
data %>% extract(1:3)
```

Your `extract_col` version is entirely redundant.

Note that base also uses the missing arguments syntax for rows:

```
#rows
data[1:3, ]
#cols
data[1:3]
```

Furthermore, having both `extract_row` and `extract` (or renamed to `extract_col`) would make it annoying to expand code. A common pattern is first writing something like:

`data %>% extract(desired_rows, )`

But then later realize we only want a few columns, in which case we can just add that argument and get:

`data %>% extract(desired_rows, desired_cols)`

Now, with your proposal, one would have to modify the function call:

```
#first
data %>% extract_rows(desired_rows)
#second
data %>% extract(desired_rows, desired_cols)
```

Unless you want to force the user to chain extracts:

```
#two part
data %>% extract_rows(desired_rows) %>% extract_cols(desired_cols)
```

I'll ask RStudio to white-list the `extract` function.

**Comment by danklotz** [comment]
Thank you for your response! 
If it is ok, I would like to respond to it. 
Your argument is only true for `data.frames` (probably tibbles and some other stuff to, but…).
I do agree that a `data.frame` is most often the way to go, so you have a point there and I would be happy if the `extract` function would be whitelisted.
 That said, I think that always chain-transforming data seems in the pipeline might also be suboptimal. If the data is, for example, a matrix ` extract(1:3)` != ` extract(,1:3)`:
```
g <- matrix(data = 1:16, ncol = 4) 
# not the same:
g %>% extract(1:3) 
g %>% extract(, 1:3)
```
I also fail to see why the original `extract` would need to be modified. There is no inherent reason to change an alias just because new (better defined) infix functions are added, or am I seeing this wrong?I mean applying this logic the `lapply` function in base-R would need to be changed as it does not provide any additional functionality over the `sapply` function…

**Comment by danklotz** [comment]
On a second tought: 

Assuming that the behaviour of` `extract` would be (made) as descirbed by @Deleetdk. Then I would still argue that the `extract(desired_rows, )` syntax is sub-optimal. 
It produces confusion about the way functions are adressed,'(x, )', and data is adressed, [x, ] at makes the code yet again harder to read. 
 In that case I think it would actualy be better to not promote the extract function at all and use '%>% .[x, ]' instead, since it is at least somewhat conistent with the rest of R. 
But maybe I am the only one thinking like that. In that case the issue can obviously be closed and forgotten.

**Comment by chrisschuerz** [comment]
I agree with @danklotz. The most concerning issue is the ambiguity of how `extract()` treats different data types differently. One cannot assume to use the `magrittr` package exclusively with `data.frame` type. 
I do not see the great additional effort of  changing from `data %>% extract_rows(desired_rows)` to `data %>% extract(desired_rows, desired_cols)` in case of changes in the code either. 
I see however that the code might become much more precise in its interpretation when using:
```
data %>% extract_rows(desired_rows)
#for specific extraction of rows
data %>% extract_cols(desired_cols)
#for specific extraction of cols and
data %>% extract(desired_rows, desired_cols)
#if a specific subset of the data is extracted
```
This is by the way the same convention applied to `set_names()` for naming. Consequently adding these functions would bring more consistency to the whole set of alias functions.

**Comment by Deleetdk** [comment]
You do realize that `extract` and `extract2` are just synonyms for `[` and `[[`, right? The point of these is to make it easier to use the base functions in a pipeline.

One can use `%>% .[1:3, ]`, but this is not easy to read and the main point of pipes is literate programming.

> It produces confusion about the way functions are addressed,'(x, )', and data is addressed, [x, ] at makes the code yet again harder to read.

The `[`, `[[` and `$` subsetting is just syntactic candy. In reality, these are just functions which are called using () as are all other functions:

```
> iris[1:3, 1:3]
  Sepal.Length Sepal.Width Petal.Length
1          5.1         3.5          1.4
2          4.9         3.0          1.4
3          4.7         3.2          1.3
> `[`(iris, 1:3, 1:3)
  Sepal.Length Sepal.Width Petal.Length
1          5.1         3.5          1.4
2          4.9         3.0          1.4
3          4.7         3.2          1.3
```

> That said, I think that always chain-transforming data seems in the pipeline might also be suboptimal. If the data is, for example, a matrix `extract(1:3) != extract(,1:3)`.

It is common for base functions to work different for different objects. That's what generic functions do. Since `extract` is `[`, it has the same confusing functionality:

```
> iris_m = as.matrix(iris)
> iris_m[1:3]
[1] "5.1" "4.9" "4.7"
> iris_m[, 1:3]
       Sepal.Length Sepal.Width Petal.Length
  [1,] "5.1"        "3.5"       "1.4"       
  [2,] "4.9"        "3.0"       "1.4"       
  [3,] "4.7"        "3.2"       "1.3"       
  [4,] "4.6"        "3.1"       "1.5"       
  [5,] "5.0"        "3.6"       "1.4"       
  [6,] "5.4"        "3.9"       "1.7"       
```

As for dim names, for data frames `colnames` and `names` refer to the same thing. So `set_colnames` is redundant, but `set_rownames` is necessary. The is easy to see if one knows that data frames are just lists and the names refer to the elements of the list. For other objects, `names` and `colnames` do not do the same. Using `colnames<-` on an atomic gives an error.

IMO not worth implementing an `extract_rows`. But if you want, here's one you can use:

```
extract_rows = function(x, i) {
  if (is.data.frame(x)) return(x[i, , drop = FALSE])
  if (is.matrix(x)) return(x[i, , drop = FALSE])
  stop("x was neither a data frame or matrix")
}
```

```
> iris %>%
+ extract_rows(2)
  Sepal.Length Sepal.Width Petal.Length Petal.Width Species
2          4.9           3          1.4         0.2  setosa
> iris %>%
+ as.matrix %>%
+ extract_rows(2)
     Sepal.Length Sepal.Width Petal.Length Petal.Width Species 
[1,] "4.9"        "3.0"       "1.4"        "0.2"       "setosa"
> 1:3 %>%
+ extract_rows(2)
 Error in extract_rows(., 2) : x was neither a data frame or matrix 
```

Please note: I'm not the package developer.
